### ISSUE TITLE
Sketchy behaviour shelling out to system commands

### ISSUE THREAD

**Issue by m4rkw** [ProtonMail/proton-bridge#257]
Hello,

proton-bridge seems to shell out to a bunch of system commands that it really should have no business executing and looks sketchy as hell when profiling it for apparmor.

So far I've seen it shell out to: git, pass, bash, which, getopt, tree, tail, sed, dirname, mkdir, gpg, base64, rm, rmdir, gpg-agent and lsof.

Why the hell does an IMAP and SMTP bridge need to shell out to lsof? A sane apparmor profile for proton-bridge would not include access to any of this but because it appears to have been lazily programmed to shell out rather than write functions in its native language this is now necessary in order to run it.

Mark

````
Jan 30 12:12:49 d kernel: [45532.229736] audit: type=1400 audit(1643544769.318:1283): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/git" pid=593346 comm="proton-bridge" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:12:49 d kernel: [45532.262449] audit: type=1400 audit(1643544769.354:1284): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/pass" pid=593347 comm="proton-bridge" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:13:33 d kernel: [45576.003257] audit: type=1400 audit(1643544813.094:1287): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/bash" pid=594111 comm="pass" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:13:55 d kernel: [45598.270244] audit: type=1400 audit(1643544835.350:1296): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/which" pid=594323 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:13:55 d kernel: [45598.282082] audit: type=1400 audit(1643544835.358:1298): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/getopt" pid=594324 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:13 d kernel: [45616.501938] audit: type=1400 audit(1643544853.590:1301): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/tree" pid=594843 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:13 d kernel: [45616.522749] audit: type=1400 audit(1643544853.594:1303): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/tail" pid=594844 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:13 d kernel: [45616.535051] audit: type=1400 audit(1643544853.594:1305): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/sed" pid=594845 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:13 d kernel: [45616.549587] audit: type=1400 audit(1643544853.630:1307): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/dirname" pid=594853 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:13 d kernel: [45616.561744] audit: type=1400 audit(1643544853.630:1309): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/mkdir" pid=594854 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:14:27 d kernel: [45630.100045] audit: type=1400 audit(1643544867.178:1321): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/gpg" pid=595029 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:16:17 d kernel: [45740.856451] audit: type=1400 audit(1643544977.946:1367): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/base64" pid=597042 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:17:41 d kernel: [45823.972078] audit: type=1400 audit(1643545061.062:1382): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/rm" pid=598762 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 12:17:41 d kernel: [45823.994518] audit: type=1400 audit(1643545061.086:1384): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/rmdir" pid=598765 comm="bash" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Jan 30 13:15:06 d kernel: [  221.275738] audit: type=1400 audit(1643548506.311:30): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/gpg-agent" pid=4601 comm="gpg" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
Feb  3 17:51:44 d kernel: [336337.210188] audit: type=1400 audit(1643910704.755:36): apparmor="DENIED" operation="exec" profile="/usr/local/bin/proton-bridge" name="/usr/bin/lsof" pid=229716 comm="proton-bridge" requested_mask="x" denied_mask="x" fsuid=1006 ouid=0
````

**Comment by doggodanubus** [comment]
Thanks Mark, I just entered a ticked as I can't compile or deploy the bridge software again. I'm not liking this at all as I won't be able to use this service I paid for as webmail is vile and I won't run an app I cannot at least compile and store the code for.

**Comment by kortschak** [comment]
Note that the exec'ing of bash to change resource limits could easily use [`unix.Getrlimit`](https://pkg.go.dev/golang.org/x/sys/unix#Getrlimit) on linux and darwin with [less code](https://go.dev/play/p/sp2fxMV5F5A) than is currently used to shell out. The use of `lsof` is just weird.

In the tests where `diff` is exec'd I'll suggest using [`cmp.Diff`](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) or perhaps [`diff.Text`](https://pkg.go.dev/github.com/pkg/diff#Text). Something similar to the former may end up in the standard library at some point (see golang.org/issue/45200).

**Comment by m4rkw** [comment]
I would also just like to gently point out that this is not a good look for a company that is ostensibly all about security. This kind of sloppy programming could undermine trust in the service.

**Comment by andrzejsza** [comment]
indeed. this has been overlooked and needs investigation.

**Comment by kortschak** [comment]
Advice for the number of open file descriptors would be to glean the information from /proc/\<pid\>/fd using [os.ReadDir](https://go.dev/play/p/9L7LYZDvTTP) on linux. There is presumably an equivalent filesystem on darwin.

**Comment by tomtastic** [comment]
Also, `lsof -p` on Darwin will include lines for `program text (code and data)` (frameworks, etc), which will greatly skew the results if all that was wanted was the number of open file descriptors.

Better would be :
```
lsof -b -w -n -P -d^txt,^mem,^cwd,^rtd -a -p
```
- `-b` avoid using kernel functions that would block.
- `-w` avoid printing warnings when avoiding kernel functions (from `-b`)
- `-n` inhibits the conversion of network numbers to host names for network files (faster)
- `-P` inhibits the conversion of port numbers to port names for network files (faster)
- `-d` exclude (`^`) FD types of `txt` (program text), `mem` (memory-mapped files), `cwd`,`rtd` (current and root directory)
- `-a` causes list selection options to be ANDed
- `-p` as normal, to specify the PID 

eg.
```
$ sw_vers
ProductName:	macOS
ProductVersion:	12.3
BuildVersion:	21E5227a
$ lsof -p $(pgrep -f 'ProtonMail Bridge') | wc -l
     206
$ lsof -b -w -n -P -d^txt,^mem,^cwd,^rtd -a -p $(pgrep -f 'ProtonMail Bridge') | wc -l
      47
```

This should improve the accuracy on Linux too, and in my testing matches closely that of `ls -l /proc/<PID>/fd`.

**Comment by kortschak** [comment]
This ignores the elephant in the room though; looking to see if you have >95% of available file descriptors and re-exec'ing is not a good way to manage resources.

https://github.com/ProtonMail/proton-bridge/blob/61e4ca58147a6d9b34fdd6f251b89a9eeeca75b0/internal/store/ulimit.go#L47-L64

https://github.com/ProtonMail/proton-bridge/blob/61e4ca58147a6d9b34fdd6f251b89a9eeeca75b0/internal/store/event_loop.go#L230-L234

But maybe this has been recognised already, since this is the only line that `bridgeEvents.RestartBridgeEvent` is ever referenced and the literal that the constant holds is only in one location, at the const decl. So it looks like this is actually just dead code shelling out fruitlessly.

**Comment by tomtastic** [comment]
In some quick testing yesterday it didn’t look like the isFdCloseToULimit code was getting called on Darwin, so yes @kortschak this lsof exec does seem to be dead code.

**Comment by andrzejsza** [comment]
Released to beta as 2.2.1, to live as 2.2.2. Thanks for the feedback. I will close the issue for now - with Bridge getting a new architecture in V3 we won't be focusing on this more at this stage.
